crates/sui-indexer-alt-framework/src/ingestion/broadcaster.rs

```diff
@@ -70,74 +85,212 @@ where
         }
 
         // Initialize ingest_hi watch channel.
-        // Start with None (no backpressure) or Some if we received initial feedback.
-        let initial_ingest_hi = subscribers_hi
+        // Start with None or Some if we received initial feedback.
+        let mut ingest_hi = subscribers_hi
             .values()
             .copied()
             .min()
             .map(|min_hi| min_hi + buffer_size as u64);
-        let (ingest_hi_watch_tx, ingest_hi_watch_rx) = watch::channel(initial_ingest_hi);
-
-        // Spawn the broadcaster task.
-        let mut broadcaster_handle = tokio::spawn(broadcast_range(
-            start_cp,
-            end_cp,
-            retry_interval,
-            concurrency,
-            ingest_hi_watch_rx.clone(),
-            client.clone(),
-            subscribers.clone(),
-            cancel.clone(),
-        ));
+        let (ingest_hi_watch_tx, ingest_hi_watch_rx) = watch::channel(ingest_hi);
 
-        loop {
-            tokio::select! {
-                _ = cancel.cancelled() => {
-                    info!("Shutdown received, stopping ingestion");
-                    break;
-                }
+        let mut watermark = start_cp;
 
-                // Subscriber watermark update
-                // docs::#regulator (see docs/content/guides/developer/advanced/custom-indexer.mdx)
-                Some((name, hi)) = ingest_hi_rx.recv() => {
-                    subscribers_hi.insert(name, hi);
+        'outer: loop {
+            if watermark > end_cp {
+                info!("Reached end of requested checkpoint range, stopping broadcaster");
+                break;
+            }
 
-                    if let Some(min_hi) = subscribers_hi.values().copied().min() {
-                        let new_ingest_hi = min_hi + buffer_size as u64;
-                        // Update the watch channel, which will notify all waiting tasks
-                        let _ = ingest_hi_watch_tx.send(Some(new_ingest_hi));
-                        info!(ingest_hi = new_ingest_hi, "Updated ingest_hi");
+            let (mut broadcaster_handle, mut ingestion_range) = (None, None);
+            let mut is_streaming = false;
+
+            if let Some(streaming_service) = &mut streaming_service {
+                if let Err(e) = streaming_service.start_streaming().await {
```

Comment:
It would make sense to have a type/value representing the active stream that you can return here on success.